src/buffer.rs

```diff
@@ -0,0 +1,113 @@
+//! Code for handling buffers to write into.
+
+#![allow(unsafe_code)]
+
+use core::mem::MaybeUninit;
+use core::slice;
+
+use crate::io::Result;
+
+/// Wrap around a buffer and do something with the pointer and length.
+///
+/// # Safety
+///
+/// The callback `f` must actually write the return value's worth of instances of `T` to the
+/// buffer.
+#[inline]
+pub(crate) unsafe fn with_buffer<
+    T: AnyBitPattern,
+    Buf: Buffer<T>,
+    F: FnOnce(*mut T, usize) -> Result<usize>
+>(mut buffer: Buf, f: F) -> Result<Buf::Result> {
+    let (ptr, cap) = buffer.as_buffer();
+    let len = f(ptr, cap)?;
+    Ok(buffer.finish(len))
+}
+
+/// A buffer that the system can write into.
+pub unsafe trait Buffer<T: AnyBitPattern>: Sized {
+    /// The result of the process operation.
+    type Result;
+
+    /// Convert this buffer into a pointer to a buffer and its capacity.
+    fn as_buffer(&mut self) -> (*mut T, usize);
+
+    /// Convert a finished buffer pointer into its result.
+    unsafe fn finish(self, len: usize) -> Self::Result;
+}
+
+unsafe impl<T: AnyBitPattern> Buffer<T> for &mut [T] {
+    type Result = usize;
+
+    #[inline]
+    fn as_buffer(&mut self) -> (*mut T, usize) {
+        (self.as_mut_ptr(), self.len())
+    }
+
+    #[inline]
+    unsafe fn finish(self, len: usize) -> Self::Result {
+        len
+    }
+}
+
+unsafe impl<'a, T: AnyBitPattern> Buffer<T> for &'a mut [MaybeUninit<T>] {
+    type Result = (&'a mut [T], &'a mut [MaybeUninit<T>]);
+
+    #[inline]
+    fn as_buffer(&mut self) -> (*mut T, usize) {
+        (self.as_mut_ptr().cast(), self.len())
+    }
+
+    #[inline]
+    unsafe fn finish(self, len: usize) -> Self::Result {
+        let (init, uninit) = self.split_at_mut(len);
+
+        // SAFETY: The user asserts that the slice is now initialized.
+        let init = slice::from_raw_parts_mut(
+            init.as_mut_ptr().cast(),
+            init.len()
+        );
+
+        (init, uninit)
+    }
+}
+
+/// Implements [`Buffer`] around the `Vec` type.
+///
+/// This implementation fills the buffer with data and sets the length.
+#[cfg(feature = "alloc")]
+unsafe impl<T: AnyBitPattern> Buffer<T> for alloc::vec::Vec<T> {
+    type Result = alloc::vec::Vec<T>;
+
+    #[inline]
+    fn as_buffer(&mut self) -> (*mut T, usize) {
+        (self.as_mut_ptr(), self.len())
```

Comment:
I think you could get away with using `self.capacity()` here rather than just `self.len()`, which would make this approach more attractive because one could pass in a `Vec::with_capacity(n)` and read into it without initializing it first, or take an existing `Vec`, `clear()` it, and then reuse the existing memory directly.

Another option would be to use `.`[`spare_capacity_mut`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.spare_capacity_mut)`().as_buffer()`, and say the behavior here is to append elements to the `Vec` rather than rewriting the whole `Vec`. That seems nice for `read_to_end` kinds of use cases. On the other hand, I wonder if that's too subtle. Given a `v: Vec<T>`, passing in `v` would append, while `&mut v` would start from the beginning. And passing `&v` to write would still start from the beginning.